### PR TITLE
File can be a remote source

### PR DESCRIPTION
**Pull request by xkraty**

### CODE REVIEW

File: lib/doc2pdf.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ module Doc2pdf
   class Error < StandardError; end
 
   # Example: "foo {bar} spam {egg} asd" -> ['bar', 'egg']
-  PLACEHOLDER_PATTERN = /{(\w*)}/.freeze
+  PLACEHOLDER_PATTERN = /({\w*})/.freeze
```
Comment (olistik) [inline lib/doc2pdf.rb:13]: Now I get why you want to capture also `{` and `}`. It's that so you can replace the whole string later.
Gotcha! Please update the comment above then.

File: lib/doc2pdf.rb [hunk 2]
```diff
@@ -21,7 +21,9 @@ def self.extract(document:)
   def self.replace!(document:, replacer:)
     DocumentTraversal.new(document: document).each do |item|
       search(item.text).each do |occurrence|
-        item.substitute(occurrence, replacer.call(key: occurrence))
+        tm = replacer.call(key: occurrence)
+        puts "Occurrence #{occurrence}, Replace #{tm}"
```
Comment (olistik) [inline lib/doc2pdf.rb:25]: Please remove debugging code.
Also, regarding `tm`, you could use a more descriptive name.

File: lib/doc2pdf.rb [hunk 3]
```diff
@@ -31,10 +33,10 @@ def self.replace!(document:, replacer:)
   def self.replace_and_save!(document:, output_base_path:, replacer:)
     FileUtils.mkdir_p(File.dirname(output_base_path))
 
-    replace!(document: document, replacer: replacer)
+    document = replace!(document: document, replacer: replacer)
 
-    doc_path = "#{output_base_path}.doc"
-    pdf_path = "#{output_base_path}.pdf"
+    doc_path = "#{output_base_path}/file.doc"
```
Comment (olistik) [inline lib/doc2pdf.rb:38]: This is an arbitrary convention that I think could pose a tricky constraint to the users.
Not to mention that if we want to do so, then we have to update the documentation accordingly.

File: lib/doc2pdf/document.rb [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
 
 require "docx"
+require "uri"
 
 module Doc2pdf
-  # Wraps a Docx::Document.
+  # Wraps a ::Docx::Document.
   class Document
-    attr_reader :docx
+    attr_accessor :docx
```
Comment (olistik) [inline lib/doc2pdf/document.rb:9]: Do you actually need the setter method `#docx=` to be public?

File: lib/doc2pdf/document.rb [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
 
     def initialize(file:)
-      self.docx = Docx::Document.open(file)
+      file = load_from_uri(file)
```
Comment (olistik) [inline lib/doc2pdf/document.rb:12]: As per my previous comment, I'd advise against this.
If you want to put this responsibility onto this gem, then we could extract the loading behaviour to a separate object but I don't think it's worth it.

**Review by olistik — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -79,10 +79,8 @@ document = Doc2pdf::Document.new(file: 'my_document.doc')
 # passing the file stream
 document = Doc2pdf::Document.new(file: File.open('my_document.doc'))
 
-# passing the stream of a remote resource
-require 'open-uri'
-
-document = Doc2pdf::Document.new(file: URI.open('https://mywebsite.com/my_document.doc'))
```
[the remark this answers is not in the document]
Comment (xkraty) [reply README.md:85]: It still a file even if remote, if file is not reachable already raise the error?!

File: lib/doc2pdf/document.rb (hunk 4 above)
Comment (xkraty) [reply lib/doc2pdf/document.rb:9]: I did commit by mistake some test changes

**Review by xkraty — COMMENTED** [review]
Gonna push changes

File: lib/doc2pdf.rb (hunk 2 above)
Comment (xkraty) [reply lib/doc2pdf.rb:25]: As before, wrong commit

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (olistik) [reply README.md:85]: Nope, it's a path, a URL. A file is a different kind of object.
Here you'd better name it `url` or even `address`. You could try to uniform it with `path` but then you'd still have the issue of the unnecessary responsibility.

File: lib/doc2pdf.rb (hunk 3 above)
Comment (xkraty) [reply lib/doc2pdf.rb:38]: It was throwing error without name, gonna add a filename param

Comment (olistik) [reply lib/doc2pdf.rb:38]: I added the instructions in the `README.md`:

Screenshot 2021-04-10 at 18 23 09

RTFM :-D

**Review by xkraty — COMMENTED** [review]
Merging for online testing

File: lib/doc2pdf.rb (hunk 3 above)
Comment (olistik) [reply lib/doc2pdf.rb]: @xkraty what about this? 😅

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (olistik) [reply README.md:85]: @xkraty This is still an issue IMO.

**Pull request merged**